src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -0,0 +1,256 @@
+use crate::{
+    error::Error,
+    ff::Field,
+    protocol::{
+        boolean::multiply_all_shares,
+        context::Context,
+        sort::MultiBitPermutationStep::{EqualityBitChecker, MultiplyAcrossBits, Sop},
+        RecordId,
+    },
+    secret_sharing::SecretSharing,
+};
+
+use futures::future::try_join_all;
+
+/// This is an implementation of `GenMultiBitSort` (Algorithm 11) described in:
+/// "An Efficient Secure Three-Party Sorting Protocol with an Honest Majority"
+/// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
+/// <https://eprint.iacr.org/2019/695.pdf>.
+///
+/// Protocol to compute a secret sharing of a permutation, after sorting on multiple bits `num_multi_bits`.
+/// At a high level, the protocol works as follows:
+/// 1. Start with a vector of list of `L*n` secret shares `[[x1_1 ... x1_n], .. , [xL_1 ... xL_n]]` where each is a secret sharing of either zero or one.
+///    Here, L is the number of multi bits which are procrssed together (`num_multi_bits`) and n is the number of records
+/// 2. Equality Bit Checker : For j in 0 to 2 pow `num_multi_bits`
+///    i. Get binary representation of j (B1 .. BL)
+///    ii. For i in `num_multi_bits`
+///      a. Locally compute `mult_inputs` as (Bi * `xi_j` + (1-Bi)(1- `xi_j`))
+///   iii. Multiply all `mult_inputs` for this j
+/// 3. Compute accumulated sum: For j in 0 to 2 pow `num_multi_bits`
+///    i. For each record
+///       a. Calculate accumulated `total_sum` = s + `mult_output`
+/// 4. Compute the final output using sum of products executed in parallel for each record.
+#[allow(dead_code)]
+pub async fn multi_bit_permutation<'a, F: Field, S: SecretSharing<F>, C: Context<F, Share = S>>(
+    ctx: C,
+    input: &[Vec<S>],
+) -> Result<Vec<S>, Error> {
+    let num_multi_bits = input.len();
+    let num_records = input[0].len();
+    let num_possible_bit_values = 2 << (num_multi_bits - 1);
+
+    // Equality bit checker: this checks if each secret shared record is equal to any of numbers between 0 and num_possible_bit_values
+    let mut equality_check_futures = Vec::with_capacity(num_possible_bit_values * num_records);
+    for j in 0..num_possible_bit_values {
+        let ctx_equality_checker = ctx.narrow(&EqualityBitChecker);
```

Comment:
I don't think this narrow call is necessary, because within get_bit_equality_checkers you're immediately calling narrow again:
let ctx_across_bits = ctx.narrow(&MultiplyAcrossBits);
So as far as I can tell, you never actually use this ctx_equality_checker context on its own.

Reply:
Since we have an async move just after this, it expects to receive a context which can be consumed by the async block. Changing it to clone the context instead (which is what others are also doing in the code)